File: backend/src/main/java/fr/sylvainjanet/tracker/journal/application/service/GetWeightMeasurementByDateService.java
package fr.sylvainjanet.tracker.journal.application.service;

import fr.sylvainjanet.tracker.journal.application.port.in.dtos.query.GetWeightMeasurementByDateQuery;
import fr.sylvainjanet.tracker.journal.application.port.in.dtos.result.GetWeightMeasurementByDateResult;
import fr.sylvainjanet.tracker.journal.application.port.in.usecase.GetWeightMeasurementByDateUseCase;
import fr.sylvainjanet.tracker.journal.application.port.out.dtos.criteria.GetWeightMeasurementByDateCriteria;
import fr.sylvainjanet.tracker.journal.application.port.out.dtos.outcome.GetWeightMeasurementByDateOutcome;
import fr.sylvainjanet.tracker.journal.application.port.out.gateway.store.WeightMeasurementStore;
import fr.sylvainjanet.tracker.journal.domain.Weight;
import fr.sylvainjanet.tracker.journal.domain.WeightMeasurement;
import java.time.LocalDate;
import java.util.Objects;
import java.util.Optional;

public final class GetWeightMeasurementByDateService implements GetWeightMeasurementByDateUseCase {

    private final WeightMeasurementStore store;

    public GetWeightMeasurementByDateService(WeightMeasurementStore store) {
        this.store = Objects.requireNonNull(store, "store must not be null");
    }

    @Override
    public Optional<GetWeightMeasurementByDateResult> get(GetWeightMeasurementByDateQuery query) {
        Objects.requireNonNull(query, "query must not be null");
        LocalDate date = queryToDomain(query);

        Optional<GetWeightMeasurementByDateOutcome> outcomeOptional =
                store.getByDate(toInstruction(date));

        Optional<WeightMeasurement> domainOptional = outcomeOptional.map(this::outcomeToDomain);

        return domainOptional.map(this::toResult);
    }

    private GetWeightMeasurementByDateResult toResult(WeightMeasurement measurement) {
        return new GetWeightMeasurementByDateResult(
                measurement.date(), measurement.weightInKilograms());
    }

    private GetWeightMeasurementByDateCriteria toInstruction(LocalDate date) {
        return new GetWeightMeasurementByDateCriteria(date);
    }

    private LocalDate queryToDomain(GetWeightMeasurementByDateQuery query) {
        return query.date();
    }

    private WeightMeasurement outcomeToDomain(GetWeightMeasurementByDateOutcome outcome) {
        return WeightMeasurement.create(outcome.date(), Weight.of(outcome.weightInKg()));
    }
}

File: frontend/src/app/contexts/journal/application/service/get-weight-measurement-by-date.service.ts
import type {
  GetWeightMeasurementByDateCriteria,
  GetWeightMeasurementByDateOutcomeData,
  WeightMeasurementStore,
} from '../port/out/weight-measurement.store';
import { type CalendarDate, calendarDate } from '../../../../shared/api/shared.calendar';
import type {
  GetWeightMeasurementByDateQuery,
  GetWeightMeasurementByDateResult,
  GetWeightMeasurementByDateResultData,
  GetWeightMeasurementByDateUseCase,
} from '../port/in/get-weight-measurement-by-date.use-case';
import { WeightMeasurement } from '../../../../shared/api/shared.weight-measurement';
import { Weight } from '../../../../shared/api/shared.weight-measurement';

export class GetWeightMeasurementByDateService implements GetWeightMeasurementByDateUseCase {
  constructor(private readonly store: WeightMeasurementStore) {}

  async get(query: GetWeightMeasurementByDateQuery): Promise<GetWeightMeasurementByDateResult> {
    if (!query) {
      throw new Error('query must not be null');
    }

    const domain = queryToDomain(query);

    const criteria = toCriteria(domain);
    const outcome = await this.store.getByDate(criteria);

    if (outcome.kind === 'failed') {
      throw new Error('Failed to get weight measurement: ' + outcome.outcomeData.errorMessage);
    }

    if (outcome.kind === 'found') {
      const domainResult = outcomeDataToDomain(outcome.outcomeData);

      return {
        kind: 'found',
        resultData: toResultData(domainResult),
      };
    }

    return { kind: 'not-found' };
  }
}

function queryToDomain(query: GetWeightMeasurementByDateQuery): CalendarDate {
  return calendarDate(query.date);
}

function outcomeDataToDomain(command: GetWeightMeasurementByDateOutcomeData): WeightMeasurement {
  return WeightMeasurement.create(calendarDate(command.date), Weight.of(command.weightInKg));
}

function toCriteria(domain: CalendarDate): GetWeightMeasurementByDateCriteria {
  return {
    date: domain,
  };
}

function toResultData(domain: WeightMeasurement): GetWeightMeasurementByDateResultData {
  return {
    date: domain.calendarDate,
    weightInKg: domain.weightInKilograms(),
  };
}
